utils/markdown.go

```diff
@@ -0,0 +1,179 @@
+				util.Prioritized(newNotificationRenderer(), 500),
+			)),
+		),
+	)
+
+	var buf bytes.Buffer
+	if err := md.Convert([]byte(markdown), &buf); err != nil {
+		return ""
+	}
+
+	return strings.TrimSpace(buf.String())
+}
+
+type notificationRenderer struct {
+}
+
+func newNotificationRenderer() *notificationRenderer {
+	return &notificationRenderer{}
+}
+
+func (r *notificationRenderer) RegisterFuncs(reg renderer.NodeRendererFuncRegisterer) {
+	// block
+	reg.Register(ast.KindDocument, r.renderDefault)
+	reg.Register(ast.KindHeading, r.renderHeading)
+	reg.Register(ast.KindBlockquote, r.renderDefault)
+	reg.Register(ast.KindCodeBlock, r.renderCodeBlock)
+	reg.Register(ast.KindFencedCodeBlock, r.renderFencedCodeBlock)
+	reg.Register(ast.KindHTMLBlock, r.renderDefault)
+	reg.Register(ast.KindList, r.renderDefault)
+	reg.Register(ast.KindListItem, r.renderListItem)
+	reg.Register(ast.KindParagraph, r.renderParagraph)
+	reg.Register(ast.KindTextBlock, r.renderTextBlock)
+	reg.Register(ast.KindThematicBreak, r.renderDefault)
+
+	// inlines
+	reg.Register(ast.KindAutoLink, r.renderDefault)
+	reg.Register(ast.KindCodeSpan, r.renderDefault)
+	reg.Register(ast.KindEmphasis, r.renderDefault)
+	reg.Register(ast.KindImage, r.renderDefault)
+	reg.Register(ast.KindLink, r.renderDefault)
+	reg.Register(ast.KindRawHTML, r.renderDefault)
+	reg.Register(ast.KindText, r.renderText)
+	reg.Register(ast.KindString, r.renderString)
+
+	// table
+	reg.Register(astExt.KindTable, r.renderTable)
+
+	// strikethrough
+	reg.Register(astExt.KindStrikethrough, r.renderDefault)
+}
+
+// renderDefault renderer function to renderDefault without changes
+func (r *notificationRenderer) renderDefault(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	return ast.WalkContinue, nil
+}
+
+func (r *notificationRenderer) renderHeading(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	if !entering {
+		if node.NextSibling() != nil {
+			_, _ = w.WriteString(" ")
+		}
+	}
+	return ast.WalkContinue, nil
+}
+
+func (r *notificationRenderer) renderListItem(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	if entering {
+		fc := node.FirstChild()
+		if fc != nil {
+			if _, ok := fc.(*ast.TextBlock); !ok {
+				_ = w.WriteByte(' ')
+			}
+		}
+	} else {
+		if node.NextSibling() != nil {
+			_, _ = w.WriteString(" ")
+		}
+	}
+	return ast.WalkContinue, nil
+}
+
+func (r *notificationRenderer) renderParagraph(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	if !entering {
+		if node.NextSibling() != nil {
+			_, _ = w.WriteString(" ")
+		}
+	}
+	return ast.WalkContinue, nil
+}
+
+func (r *notificationRenderer) renderCodeBlock(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	n := node.(*ast.CodeBlock)
+	source = bytes.ReplaceAll(source, []byte("\n"), []byte(" "))
```

Comment:
There is no need to replace `\n`. Mobile can display it correctly.

Reply:
That means, in the code block there are characters `\n` that will be handled by mobile ? If yes I will update the unit test, https://github.com/mattermost/mattermost-server/pull/17775/files#diff-fd01f7b59eaca4eb783d3811157920648fb0fc7915e5cf0b4ab6202aa8075b6bR54-R58

Reply:
Yes, they will be.